design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,468 @@
+  , system-assigned-identity, user-assigned-identity]
+  ` for `authn
+  -azure`)
+- output
+  - `constraints` hash.
+   
+#### Validate the Application Identity is defined correctly
+
+In the Azure Authenticator and in the K8s Authenticator we have requirements on 
+how the annotations are defined. As can be seen 
+[here](/app/domain/authentication/authn_azure/validate_application_identity.rb), 
+we have the following validations:
+  - Validate annotations are permitted
+    - Here we validate that the `authn-azure` annotations are known. For
+      example, the annotation `authn-azure/subscription-id` is valid, whereas
+      the annotation `authn-azure/blah` is invalid. We do that to minimize
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required constraints exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate constraint combinations
+    - For example, in the Azure Authenticator we do not allow hosts to have
+     both the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we do not allow hosts to have any combination of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+      
+Also here we can see some generalization. 
+In this case the general class will be defined as follows:
+- input
+  - The following lists of annotation names
+    - permitted
+    - required
+    - non-permitted combinations
+  - The `constraints` hash built in the previous step
+- output
+  - Nothing if validation succeeded, relevant error if it failed
+
+#### Validate the Application Identity matches the authentication request
+
+Once we extracted the Application Identity from the host's annotations, and
+verified it is defined correctly, we can validate that the authentication
+request is valid by validating the Application Identity. For example, if
+the k8s host is permitted to authenticate from namespace `namespace-1`, and
+the authentication request was sent from a pod that is in namespace
+`namespace-2` we will fail the request.
+
+In a way, we can define this step as comparing 2 objects of Application
+Identity - one that was extracted from the host and one that was extracted
+from the authentication request. In the example above we will have the
+following Application Identity objects:
+ - Host Application Identity
+   - constraints
+     - namespace: `namespace-1`
+ - Authentication request Application Identity
+   - constraints
+     - namespace: `namespace-2`
```

Comment:
where does the namespace info come from in authn-k8s when an authentication request is sent from an app with a host identity defined using annotations?

previously the host would have authenticated as `{namespace}/service_account/{service account}`; the namespace / service account would be sent with the request in the form of the host ID (`CONJUR_AUTHN_LOGIN`) and the service_id would be extracted from the request route

now when a host authenticates as `my-host` in the `CONJUR_AUTHN_LOGIN`, where the host was defined in policy using the `{service account}` and `{namespace}` values as host annotations, how does the server extract the "authentication request application identity" values for service account and namespace?

sorry if this is more clearly defined elsewhere, I've just never thought about this aspect of defining hosts by annotations before and your example is raising this question for me

Reply:
As can be seen in [our docs](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/Kubernetes_AppIdentity.htm?Highlight=application%20identity#DefineKubernetesResourcesasDAPhosts) the namespace can be defined in the annotation `authn-k8s/namespace`. 

the host `{namespace}/service_account/{service account}` can be defined as follows:
```
- host 
  id: some-host
  annotations:
    - authn-k8s/namespace: {namespace}
    - authn-k8s/service-account: {service account}
```

The extraction of the values from the request stayed the same, we didn't touch that logic.